modules/core/src/test/java/org/apache/ignite/internal/client/thin/ReliableChannelTest.java

```diff
@@ -0,0 +1,229 @@
+            .setAddressesFinder(new TestAddressFinder("127.0.0.1:8000", "127.0.0.1:8001", "127.0.0.1:8002"));
+
+        checkDoesNotReinit(ccfg);
+    }
+
+    /** */
+    private void checkDoesNotReinit(ClientConfiguration ccfg) {
+        ReliableChannel rc = new ReliableChannel(chFactory, ccfg, null);
+        rc.initConnection();
+        List<ReliableChannel.ClientChannelHolder> originalChannels = rc.getChannelHolders();
+
+        // Imitate topology change.
+        rc.initChannelHolders(true);
+        List<ReliableChannel.ClientChannelHolder> newChannels = rc.getChannelHolders();
+
+        assertSame(originalChannels, newChannels);
+        IntStream.range(0, 3).forEach(i -> {
+            assertSame(originalChannels.get(i), newChannels.get(i));
+            assertFalse(originalChannels.get(i).isClosed());
+        });
+        assertEquals(3, newChannels.size());
+    }
+
+    /** Checks that node channels are persisted if channels are reinit with static address configuration. */
+    @Test
+    public void testNodeChannelsAreNotCleaned() {
+        ClientConfiguration ccfg = new ClientConfiguration()
+            .setAddresses("127.0.0.1:8000", "127.0.0.1:8001", "127.0.0.1:8002");
+
+        ReliableChannel rc = new ReliableChannel(chFactory, ccfg, null);
+        rc.initConnection();
+        // Trigger TestClientChannel creation.
+        rc.service(null, null, null);
+
+        assertEquals(1, rc.nodeChannels.size());
+
+        // Imitate topology change.
+        rc.initChannelHolders(true);
+
+        assertEquals(1, rc.nodeChannels.size());
+    }
+
+    /** Checks that channels are changed (add new, remove old) and close channels if reinitialization performed. */
+    @Test
+    public void testDynamicAddressReinitializedCorrectly() {
+        ClientConfiguration ccfg = new ClientConfiguration()
+            .setAddressesFinder(new TestAddressFinder("127.0.0.1:8000", "127.0.0.1:8003"));
+
+        ReliableChannel rc = new ReliableChannel(chFactory, ccfg, null);
+        rc.initConnection();
+
+        List<ReliableChannel.ClientChannelHolder> originChannels = Collections.unmodifiableList(rc.getChannelHolders());
+        // Imitate topology change.
+        rc.initChannelHolders(true);
+
+        List<ReliableChannel.ClientChannelHolder> closedChannels = originChannels.stream()
+            .filter(r -> r.isClosed())
+            .collect(Collectors.toList());
+
+        assertEquals(2, closedChannels.size());
+
+        List<ReliableChannel.ClientChannelHolder> reuseChannel = originChannels.stream()
+            .filter(c -> !c.isClosed())
+            .collect(Collectors.toList());
+
+        assertEquals(1, reuseChannel.size());
+
+        List<ReliableChannel.ClientChannelHolder> newChannels = rc.getChannelHolders();
+        assertEquals(2, newChannels.size());
+        assertTrue(newChannels.get(0) == reuseChannel.get(0) || newChannels.get(1) == reuseChannel.get(0));
```

Comment:
contains invoke `equals` but I check references.

Reply:
`equals` for `ClientChannelHolder` only compare references :)

Reply:
Currently it is, but the behavior may change in future and test will start check different case (we'll be lucky enough it will fail). Also I believe that explicit check is better.